src/variational/VariationalInference.jl

```diff
@@ -0,0 +1,171 @@
+    error("Turing.Variational.grad!: unmanaged variational inference algorithm: $(typeof(alg))")
+
+"""
+    vi(model::Model, alg::VariationalInference)
+    vi(model::Model, alg::VariationalInference, q::VariationalPosterior)
+
+Constructs the variational posterior from the `model` and performs the optimization following the configuration of the given `VariationalInference` instance.
+"""
+vi(model::Model, alg::VariationalInference) = 
+    error("Turing.Variational.vi: unmanaged variational inference algorithm: $(typeof(alg))")
+vi(model::Model, alg::VariationalInference, q::VariationalPosterior) =
+    error("Turing.Variational.vi: unmanaged variational inference algorithm: $(typeof(alg))")
+
+# default implementations
+function grad!(vo, alg::VariationalInference{AD}, q::VariationalPosterior, model::Model, θ::AbstractVector{T}, out::DiffResults.MutableDiffResult, args...) where {T <: Real, AD <: ForwardDiffAD}
+    # TODO: this probably slows down executation quite a bit; exists a better way of doing this?
+    f(θ_) = - vo(alg, q, model, θ_, args...)
+
+    chunk_size = getchunksize(alg)
+    # Set chunk size and do ForwardMode.
+    chunk = ForwardDiff.Chunk(min(length(θ), chunk_size))
+    config = ForwardDiff.GradientConfig(f, θ, chunk)
+    ForwardDiff.gradient!(out, f, θ, config)
+end
+
+# TODO: implement for `Tracker`
+# function grad(vo::ELBO, alg::ADVI, q::MeanField, model::Model, f, autodiff::Val{:backward})
+#     vo_tracked, vo_pullback = Tracker.forward()
+# end
+function grad!(vo, alg::VariationalInference{AD}, q::VariationalPosterior, model::Model, θ::AbstractVector{T}, out::DiffResults.MutableDiffResult, args...) where {T <: Real, AD <: TrackerAD}
+    θ_tracked = Tracker.param(θ)
+    y = - vo(alg, q, model, θ_tracked, args...)
+    Tracker.back!(y, 1.0)
+
+    DiffResults.value!(out, Tracker.data(y))
+    DiffResults.gradient!(out, Tracker.grad(θ_tracked))
+end
+
+"""
+    optimize!(vo, alg::VariationalInference{AD}, q::VariationalPosterior, model::Model, θ; optimizer = ADAGrad())
+
+Iteratively updates parameters by calling `grad!` and using the given `optimizer` to compute the steps.
+"""
+function optimize!(vo, alg::VariationalInference{AD}, q::VariationalPosterior, model::Model, θ; optimizer = ADAGrad()) where AD
+    # TODO: this default constructor for `optimizer` is probably dangerous because of the `acc` field being an `IdDict`!!!
+
+    # TODO: should we always assume `samples_per_step` and `max_iters` for all algos?
+    alg_name = alg_str(alg)
+    samples_per_step = alg.samples_per_step
+    max_iters = alg.max_iters
+
+    # number of previous gradients to use to compute `s` in adaGrad
+    stepsize_num_prev = 10
+    
+    # setup
+    # var_info = Turing.VarInfo()
+    # model(var_info, Turing.SampleFromUniform())
+    # num_params = size(var_info.vals, 1)
+    num_params = length(q)
+
+    # # buffer
+    # θ = zeros(2 * num_params)
+
+    # HACK: re-use previous gradient `acc` if equal in value
```

Comment:
What does this mean?

Reply:
Multiple of the optimisers have an `IdDict` attached to them as they need to track certain quantities. Therefore, if the user performs VI by repeatedly calling `optimize!` but is for some reason not using the same `θ` ("same" in the sense that it has the same `Id` not just value), you end up creating a different quantity in the optimizer for each call. In this case the algorithm is probably not doing what the user _thinks_ it's doing.